File: clients/blazor-admin/Services/UiLanguageCode.cs
using System.Text.RegularExpressions;

namespace Blazor.Admin.Services;

/// <summary>
/// UI and API use culture codes (e.g. tr-TR, en-US). Legacy localStorage may contain display names like "Türkçe" from older clients.
/// </summary>
public static class UiLanguageCode
{
    private static readonly Regex CultureLike = new(
        @"^[a-zA-Z]{2,3}(-[a-zA-Z0-9]{2,12})+$",
        RegexOptions.CultureInvariant | RegexOptions.Compiled);

    /// <summary>
    /// Invalid tags that match BCP 47 shape but use non-flag region codes (e.g. en-EN: "EN" is not ISO 3166-1).
    /// </summary>
    private static string CanonicalizeCultureCode(string t)
    {
        var segments = t.Split(['-', '_'], StringSplitOptions.RemoveEmptyEntries);
        if (segments.Length >= 2
            && string.Equals(segments[0], "en", StringComparison.OrdinalIgnoreCase)
            && string.Equals(segments[1], "en", StringComparison.OrdinalIgnoreCase))
        {
            return "en-US";
        }

        return t;
    }

    public static string Normalize(string? raw)
    {
        if (string.IsNullOrWhiteSpace(raw))
        {
            return "tr-TR";
        }

        var t = raw.Trim();
        if (CultureLike.IsMatch(t))
        {
            return CanonicalizeCultureCode(t);
        }

        return t switch
        {
            "Türkçe" => "tr-TR",
            "Turkish" => "tr-TR",
            "English" => "en-US",
            "İngilizce" => "en-US",
            "Ingilizce" => "en-US",
            _ => "tr-TR"
        };
    }
}

File: clients/react-admin/src/core/ui-language-code.ts
/**
 * UI and API use culture codes (e.g. tr-TR). /languages/codes returns SelectionItem: id = Language.Code, label = Language.Name.
 * Legacy localStorage may contain display names like "Türkçe".
 */

const cultureLike = /^[a-zA-Z]{2,3}(-[a-zA-Z0-9]{2,12})+$/;

function canonicalizeCultureCode(t: string): string {
  const segments = t.split(/[-_]/);
  if (segments.length >= 2 && segments[0].toLowerCase() === "en" && segments[1].toLowerCase() === "en") {
    return "en-US";
  }
  return t;
}

export function normalizeUiLanguage(raw: string | null | undefined): string {
  if (raw == null || String(raw).trim() === "") {
    return "tr-TR";
  }
  const t = String(raw).trim();
  if (cultureLike.test(t)) {
    return canonicalizeCultureCode(t);
  }
  switch (t) {
    case "Türkçe":
      return "tr-TR";
    case "Turkish":
      return "tr-TR";
    case "English":
      return "en-US";
    case "İngilizce":
    case "Ingilizce":
      return "en-US";
    default:
      return "tr-TR";
  }
}

/** Map API language lookup row to culture code; never use label (display name). */
export function cultureCodeFromLanguageLookupRow(item: Record<string, unknown>): string | null {
  const tryKeys = ["code", "Code", "id", "Id"];
  for (const key of tryKeys) {
    const v = item[key];
    if (v == null) {
      continue;
    }
    const s = String(v).trim();
    if (s !== "") {
      return s;
    }
  }
  return null;
}
